Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli

```diff
@@ -225,26 +252,112 @@ void GetDebugDetailSurface(inout DetailSurface surface, uint material1, uint mat
     surface.m_height = 0.5;
 }
 
-bool GetDetailSurface(inout DetailSurface surface, float2 idUv, float2 uv)
+//Blend a single detail material sample (with two possible material ids) onto a DetailSurface.
+void BlendDetailMaterial(inout DetailSurface surface, uint material1, uint material2, float blend, float2 detailUv, float weight)
 {
-    uint4 material1 = TerrainSrg::m_detailMaterialIdImage.GatherRed(TerrainSrg::DetailSampler, idUv, 0).xyzw;
-    uint4 material2 = TerrainSrg::m_detailMaterialIdImage.GatherGreen(TerrainSrg::DetailSampler, idUv, 0).xyzw;
+    DetailSurface tempSurface;
+    GetDetailSurfaceForMaterial(tempSurface, material1, detailUv);
+    if (material2 != 0xFF)
+    {
+        WeightDetailSurface(tempSurface, weight * blend);
+        AddDetailSurface(surface, tempSurface);
 
-    const float maxBlendAmount = 0xFF;
-    // convert integer of 0-255 to float of 0-1.
-    float4 blends = float4(TerrainSrg::m_detailMaterialIdImage.GatherBlue(TerrainSrg::DetailSampler, idUv, 0).xyzw) / maxBlendAmount;
+        GetDetailSurfaceForMaterial(tempSurface, material2, detailUv);
+        WeightDetailSurface(tempSurface, weight * (1.0 - blend));
+        AddDetailSurface(surface, tempSurface);
+    }
+    else
+    {
+        WeightDetailSurface(tempSurface, weight);
+        AddDetailSurface(surface, tempSurface);
+    }
+}
 
+/*
+Populates a DetailSurface with material data gathered form the 4 nearest samples to detailMaterialIdUv. The weight
+of each detail material's contribution is calculated based on the distance to the center point for that sample (for
+instance, if detailMaterialIdUv falls perfectly inbetween all 4 samples, then each sample will be weighed at 25%).
+Each sample can have two different detail materials defined with a blend value to determine their relative contribution.
+The detailUv is used for samping the textures of each detail material.
+*/
+bool GetDetailSurface(inout DetailSurface surface, float2 detailMaterialIdUv, float2 detailUv)
+{
+    float2 textureSize;
+    TerrainSrg::m_detailMaterialIdImage.GetDimensions(textureSize.x, textureSize.y);
+
+    float2 detailMaterialIdCoord = detailMaterialIdUv * textureSize + 0.5;
+    float2 detailMaterailIdUv = detailMaterialIdCoord;
+    float2 gatherWeight = frac(detailMaterialIdCoord);
+
+    // Adjust the gather weight for better interpolation. This helps avoid diamond-shaped artifacts in binlinear filtering.
+    gatherWeight = 3.0 * gatherWeight * gatherWeight - 2.0 * gatherWeight * gatherWeight * gatherWeight;
+
+    // Using Load() to gather the nearest 4 samples (Gather4() isn't used because of precision issues with uvs).
+    uint4 s1 = TerrainSrg::m_detailMaterialIdImage.Load(int3((detailMaterialIdCoord + int2(0, 1) + textureSize) % textureSize, 0));
```

Comment:
Is textureSize always a power of 2? Im wondering if you can replace it with bit ops since the compiler wouldn't know the size constraints at this point.

Reply:
Currently the detail texture size is a fixed power of 2 but I'm not sure yet if it will always be that way so I want to keep the logic simple for now.